File: spring-reactive-02/learn-java/src/main/java/boki/learnjava/v1_newsequence/g_generate/GenerateEx3.java
package boki.learnjava.v1_newsequence.g_generate;

import boki.learnjava.common.SampleData;
import boki.learnjava.utils.Logger;
import reactor.core.publisher.Flux;
import reactor.util.function.Tuple2;

import java.util.Map;

/**
 * generate 개념 이해 예제
 * - 2016년도 이후의 해당 연도(2017 - 2021년)의 BTC 최고가 금액을 출력
 */
public class GenerateEx3 {
    public static void main(String[] args) {
        Map<Integer, Tuple2<Integer, Long>> map = SampleData.getBtcTopPricesPerYearMap();

        Flux
            .generate(() -> 2017, (state, sink) -> {
                if (state > 2021) {
                    sink.complete();
                }
                else {
                    sink.next(map.get(state));
                }
                return ++state;
            })
            .subscribe(Logger::onNext);
    }
}

File: spring-reactive-02/learn-kt/src/main/kotlin/boki/learnkt/v1_newsequence/g_generate/GenerateEx3.kt
package boki.learnkt.v1_newsequence.g_generate

import boki.learnjava.common.SampleData
import boki.learnkt.util.Logger
import reactor.core.publisher.Flux

/**
 * generate 개념 이해 예제
 * - 2016년도 이후의 해당 연도(2017 - 2021년)의 BTC 최고가 금액을 출력
 */
fun main() {
    val map = SampleData.getBtcTopPricesPerYearMap()

    Flux
        .generate(
            { 2017 },
            { state, sink ->
                if (state > 2021) {
                    sink.complete()
                }
                else {
                    sink.next(map[state])
                }
                return@generate state + 1
            }
        )
        .subscribe { Logger.onNext(it) }
}